controllers/gatekeeper_controller.go

```diff
@@ -207,53 +211,129 @@ func (r *GatekeeperReconciler) deployGatekeeperResources(gatekeeper *operatorv1a
 		}
 	}
 
-	for _, a := range applyAssets {
-		// Handle special cases in switch below.
-		switch {
-		case a == NamespaceFile && !r.isOpenShift():
-			// Ignore the namespace resource on Kubernetes as we default to use
-			// the same namespace as the operator, which by definition is
-			// already created as a result of executing this code.
-			continue
-		case a == RoleFile && r.isOpenShift():
-			a = openshiftAssetsDir + a
+	for _, asset := range applyAssets {
+		err := r.applyAsset(gatekeeper, asset)
+		if err != nil {
+			return err
 		}
+	}
 
-		obj, err := util.GetManifestObject(a)
+	err := r.validateWebhookDeployment()
+	if err != nil {
+		return err
+	}
+	for _, asset := range webhookAssets {
+		err := r.applyAsset(gatekeeper, asset)
 		if err != nil {
 			return err
 		}
-		if err = crOverrides(gatekeeper, a, obj, r.Namespace, r.isOpenShift()); err != nil {
+	}
+
+	return nil
+}
+
+func (r *GatekeeperReconciler) applyAsset(gatekeeper *operatorv1alpha1.Gatekeeper, asset string) error {
+	// Handle special cases in switch below.
+	switch {
+	case asset == NamespaceFile && !r.isOpenShift():
+		// Ignore the namespace resource on Kubernetes as we default to use
+		// the same namespace as the operator, which by definition is
+		// already created as a result of executing this code.
+		return nil
+	case asset == RoleFile && r.isOpenShift():
+		asset = openshiftAssetsDir + asset
+	}
+
+	obj, err := util.GetManifestObject(asset)
+	if err != nil {
+		return err
+	}
+	if err = crOverrides(gatekeeper, asset, obj, r.Namespace, r.isOpenShift()); err != nil {
+		return err
+	}
+
+	if err = r.crudResource(obj, gatekeeper, apply); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *GatekeeperReconciler) validateWebhookDeployment() error {
+	r.Log.Info(fmt.Sprintf("Validating %s deployment status", WebhookFile))
+
+	ctx := context.Background()
+	obj, err := util.GetManifestObject(WebhookFile)
+	if err != nil {
+		return err
+	}
+	deployment := &unstructured.Unstructured{}
+	deployment.SetAPIVersion(obj.GetAPIVersion())
+	deployment.SetKind(obj.GetKind())
+	namespacedName := types.NamespacedName{
+		Namespace: obj.GetNamespace(),
+		Name:      obj.GetName(),
+	}
+
+	for start := time.Now(); time.Since(start) < time.Minute; {
```

Comment:
I recommend using the [wait package in apimachinery](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait) to help simplify this logic.

Alternatively, we can return `ctrl.Result{}`: https://github.com/gatekeeper/gatekeeper-operator/blob/f692033e950e937e21ae133e819e7098e48c09da/controllers/gatekeeper_controller.go#L176 with the `RequeueAfter` field set. Another reconcile shouldn't hurt things as nothing likely changed, but if it did we can pick it up as well. If we would be concerned about making unnecessary updates, we can always check if each object has changed  since we last reconciled. and if not, skip the update.

Reply:
I should add that the wait package in apimachinery is already vendored in.

Reply:
code removed, switched to requeue